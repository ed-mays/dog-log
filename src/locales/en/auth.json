{
  "logout": "Log out",
  "authentication": "Authentication",
  "google": {
    "continueWithGoogle": "Continue with Google"
  }
}
