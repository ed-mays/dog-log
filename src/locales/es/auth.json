{
  "logout": "Cerrar sesión",
  "authentication": "Autenticación",
  "google": {
    "continueWithGoogle": "Continuar con Google"
  }
}
