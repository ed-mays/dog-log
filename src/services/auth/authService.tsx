import {
  GoogleAuthProvider,
  browserLocalPersistence,
  onAuthStateChanged,
  setPersistence,
  signInWithPopup,
  signOut as fbSignOut,
  type User as FirebaseUser,
} from 'firebase/auth';
import { auth } from '@/firebase';

export type AppUser = {
  uid: string;
  displayName: string | null;
  email: string | null;
  photoURL: string | null;
};

export type Unsubscribe = () => void;

const provider = new GoogleAuthProvider();

export async function ensurePersistence(): Promise<void> {
  // Persist sessions across reloads; safe to call multiple times.
  await setPersistence(auth, browserLocalPersistence);
}

export function mapUser(user: FirebaseUser | null): AppUser | null {
  if (!user) return null;
  return {
    uid: user.uid,
    displayName: user.displayName ?? null,
    email: user.email ?? null,
    photoURL: user.photoURL ?? null,
  };
}

export async function signInWithGoogle(): Promise<AppUser> {
  await ensurePersistence();
  try {
    const result = await signInWithPopup(auth, provider);
    // Minimal telemetry (no PII):
    console.info('[auth] signInWithGoogle success');
    return mapUser(result.user)!;
  } catch (e) {
    console.warn('[auth] signInWithGoogle failed');
    throw e;
  }
}

export async function signOut(): Promise<void> {
  try {
    await fbSignOut(auth);
    console.info('[auth] signOut success');
  } catch (e) {
    console.warn('[auth] signOut failed');
    throw e;
  }
}

export function subscribeToAuth(
  cb: (user: AppUser | null) => void,
  onError?: (error: unknown) => void
): Unsubscribe {
  // Caller can show a loading state while the first event is received.
  return onAuthStateChanged(
    auth,
    (user) => {
      console.debug('[auth] onAuthStateChanged event');
      cb(mapUser(user));
    },
    (err) => {
      console.warn('[auth] onAuthStateChanged error');
      onError?.(err);
    }
  );
}
