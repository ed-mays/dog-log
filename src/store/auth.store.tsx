import { create } from 'zustand';
import type { AppUser, Unsubscribe } from '@/services/auth/authService';
import {
  signInWithGoogle as serviceSignInWithGoogle,
  signOut as serviceSignOut,
  subscribeToAuth,
} from '@/services/auth/authService';

export type AuthState = {
  user: AppUser | null;
  initializing: boolean;
  error: unknown | null;
  initAuthListener: () => void;
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
};

let unsubscribe: Unsubscribe | null = null;

export const useAuthStore = create<AuthState>((set) => ({
  user: null,
  initializing: true,
  error: null,
  initAuthListener: () => {
    if (unsubscribe) return; // idempotent
    unsubscribe = subscribeToAuth(
      (user) => set({ user, initializing: false, error: null }),
      (error) => set({ error, initializing: false })
    );
  },
  signInWithGoogle: async () => {
    try {
      await serviceSignInWithGoogle();
      // user will be set by the listener
    } catch (error) {
      set({ error });
      throw error;
    }
  },
  signOut: async () => {
    try {
      await serviceSignOut();
      // user will be set to null by the listener
    } catch (error) {
      set({ error });
      throw error;
    }
  },
}));

export const useAuthUser = () => useAuthStore((s) => s.user);
export const useAuthStatus = () =>
  useAuthStore((s) => ({ initializing: s.initializing, error: s.error }));
