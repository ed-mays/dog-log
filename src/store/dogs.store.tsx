import { create } from 'zustand';
// Import the Dog type from your DogList module to keep type definitions DRY
import type { Dog } from '@components/DogList';

interface DogsState {
  dogs: Dog[];
  loading: boolean;
  error: string | null;
  fetchDogs: () => Promise<void>;
}

export const useDogsStore = create<DogsState>((set) => ({
  dogs: [],
  loading: false,
  error: null,
  fetchDogs: async () => {
    set({ loading: true, error: null });
    try {
      // Mocked async fetch, swap with Firebase logic later
      const mockDogs: Dog[] = [
        { id: '1', name: 'Fido', breed: 'Labrador' },
        { id: '2', name: 'Bella', breed: 'Beagle' },
      ];
      await new Promise((res) => setTimeout(res, 100));
      set({ dogs: mockDogs, loading: false });
    } catch {
      set({ error: 'Failed to load dogs.', loading: false });
    }
  },
}));
