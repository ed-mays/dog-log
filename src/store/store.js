import { create } from 'zustand';
export const useAppStore = create((set) => ({
    count: 0,
    increment: () => set((state) => ({ count: state.count + 1 })),
    decrement: () => set((state) => ({ count: state.count - 1 })),
}));
