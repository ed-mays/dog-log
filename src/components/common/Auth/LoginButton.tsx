import React from 'react';
import { useTranslation } from 'react-i18next';
import { useAuthStore } from '@store/auth.store';

type Props = {
  className?: string;
  disabled?: boolean;
};

const LoginButton: React.FC<Props> = ({ className, disabled }) => {
  const { t } = useTranslation('common');
  const signIn = useAuthStore((s) => s.signInWithGoogle);
  const initializing = useAuthStore((s) => s.initializing);

  const onClick = async () => {
    try {
      await signIn();
    } catch {
      // Error state is handled by the auth store and rendered by consumers.
    }
  };

  return (
    <button
      type="button"
      className={className}
      onClick={onClick}
      disabled={disabled || initializing}
      aria-busy={initializing || undefined}
    >
      {t('continueWithGoogle', 'Continue with Google')}
    </button>
  );
};

export default LoginButton;
