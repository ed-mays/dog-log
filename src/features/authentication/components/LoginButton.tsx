import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useAuthStore } from '@store/auth.store.tsx';
import { loadNamespace } from '@i18n';
import { Button } from '@mui/material';

export type LoginButtonProps = {
  className?: string;
  disabled?: boolean;
};

const LoginButton: React.FC<LoginButtonProps> = ({ className, disabled }) => {
  const [nsReady, setNsReady] = useState(false);
  useEffect(() => {
    let mounted = true;
    Promise.all([loadNamespace('common'), loadNamespace('auth')]).then(() => {
      if (mounted) setNsReady(true);
    });
    return () => {
      mounted = false;
    };
  }, []);

  const { t } = useTranslation();
  const signIn = useAuthStore((s) => s.signInWithGoogle);
  const initializing = useAuthStore((s) => s.initializing);
  if (!nsReady) return null;

  const onClick = async () => {
    try {
      await signIn();
    } catch {
      // Error state is handled by the auth store and rendered by consumers.
    }
  };

  return (
    <Button
      type="button"
      variant="contained"
      className={className}
      onClick={onClick}
      disabled={disabled || initializing}
      aria-busy={initializing || undefined}
      aria-label={t('continueWithGoogle', 'Continue with Google')}
      data-testid="login-button"
    >
      {t('google.continueWithGoogle', 'Continue with Google Default', {
        ns: 'auth',
      })}
    </Button>
  );
};

export default LoginButton;

/* TODO: ADDITIONAL CLEANUP
- Rename to GoogleLoginButton?
- Create Auth i18n namespace and integrate with app
- I don't like the empty catch block, I need to understand why that is that way

 */
