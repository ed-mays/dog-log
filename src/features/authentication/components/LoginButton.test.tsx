import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@test-utils';
import userEvent from '@testing-library/user-event';
import LoginButton from './LoginButton';
import { useAuthStore } from '@store/auth.store';
import { AuthState } from '@store/auth.store';
import { Suspense } from 'react';
import { waitFor } from '@testing-library/react';

describe('LoginButton', () => {
  let signInMock: ReturnType<typeof vi.fn>;
  beforeEach(() => {
    signInMock = vi.fn().mockResolvedValue(undefined);
    useAuthStore.setState((prev: AuthState) => ({
      ...prev,
      initializing: false,
      signInWithGoogle: signInMock,
    }));
  });

  it('calls signInWithGoogle on click', async () => {
    render(<LoginButton />);
    const btn = await screen.findByRole('button', {
      dataTestId: 'login-button',
    });
    await userEvent.click(btn);
    expect(signInMock).toHaveBeenCalledTimes(1);
  });

  describe('Initialization behavior', () => {
    it('is disabled while initializing', async () => {
      useAuthStore.setState((prev) => ({ ...prev, initializing: true }));
      const { asFragment, findByTestId } = render(
        <Suspense fallback={<div />}>
          <LoginButton />
          );
        </Suspense>
      );

      await waitFor(() => findByTestId('login-button'));
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('is enabled after initializing', async () => {
    useAuthStore.setState((prev) => ({ ...prev, initializing: true }));
    const { asFragment, findByTestId } = render(
      <Suspense fallback={<div />}>
        <LoginButton />
      </Suspense>
    );

    await waitFor(() => findByTestId('login-button'));
    useAuthStore.setState((prev) => ({ ...prev, initializing: false }));

    expect(asFragment()).toMatchSnapshot();
  });

  describe('i18N behavior', () => {
    it.todo('loads translations from the i18n provider', async () => {});
    it.todo('translates the button test in English', async () => {});
    it.todo('translates the button test in Spanish', async () => {});
  });
});
