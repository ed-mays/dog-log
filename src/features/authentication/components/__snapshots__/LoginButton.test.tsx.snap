// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginButton > Initialization behavior > is disabled while initializing 1`] = `
<DocumentFragment>
  <button
    aria-busy="true"
    aria-label="Continue with Google"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    data-testid="login-button"
    disabled=""
    tabindex="-1"
    type="button"
  >
    Continue with Google
  </button>
  );
</DocumentFragment>
`;

exports[`LoginButton > is disabled while initializing 1`] = `
<DocumentFragment>
  <button
    aria-busy="true"
    aria-label="Continue with Google"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    data-testid="login-button"
    disabled=""
    tabindex="-1"
    type="button"
  >
    Continue with Google
  </button>
  );
</DocumentFragment>
`;

exports[`LoginButton > is enabled after initializing 1`] = `
<DocumentFragment>
  <button
    aria-busy="true"
    aria-label="Continue with Google"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    data-testid="login-button"
    disabled=""
    tabindex="-1"
    type="button"
  >
    Continue with Google
  </button>
</DocumentFragment>
`;

exports[`is enabled after initializing 1`] = `
<DocumentFragment>
  <button
    aria-busy="true"
    aria-label="Continue with Google"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    data-testid="login-button"
    disabled=""
    tabindex="-1"
    type="button"
  >
    Continue with Google
  </button>
</DocumentFragment>
`;
